content/blog/open-source-monitoring-at-the-hpe-customer-technology-center-böblingen.md

```diff
@@ -0,0 +1,377 @@
+
+## g﻿et\_server\_urls
+
+The get\_server_urls routine is using the list of ILO IP addresses to gather the correct URLs out of the Redfish resource directory. If the ILO is alive  a Redfish client object for this ILO is created and the redfish resource directory is read. The client object is stored together with the URLs for the thermal, power and computer system information in the python directory server\_urls, which is the return value of the routine.
+
+```python
+def get_server_urls( login_account, login_password, server, lfile):
+
+    server_urls={}
+    if LINUX:
+        cmd='ping -c 2 '
+    log=logopen(lfile)
+    logwriter(log,'get_server_urls')
+    for s in server:
+        ilo_url="https://"+s['ilo']
+        s["url"]=ilo_url
+        if LINUX:
+            response=os.system(cmd+s['ilo'])   # works on Linux without root priviliges
+        else:
+            response = ping(address=s['ilo'],count=1)  # requires root priviliges on Linux
+            if(response.is_alive):
+                response = 0
+            else:
+                response = 256
+        if (response == 0):        
+            try:
+                # Create a Redfish client object
+                REDFISHOBJ = redfish.redfish_client(base_url=ilo_url, username=login_account, password=login_password)  
+                # Login with the Redfish client
+                REDFISHOBJ.login()
+                s["redfish"]=REDFISHOBJ
+                resource_instances = get_resource_directory(REDFISHOBJ, lfile)
+                for instance in resource_instances:
+                    if '#ComputerSystem.' in instance ['@odata.type']:
+                        s["ComputerSystem"]=instance['@odata.id'] 
+                    if '#Power.' in instance ['@odata.type']:
+                        s["Power"]=instance['@odata.id'] 
+                    if '#Thermal.' in instance ['@odata.type']:
+                        s["Thermal"]=instance['@odata.id']
+                if len(s) > 4:
+                    server_urls[s['ilo']]=s
+                    logwriter(log,s['ilo']+' completed')
+            except Exception as ex:
+                logwriter(log,'Exception - get_server_urls: '+s['ilo'])
+                logwriter(log,str(ex.__context__))
+                if len(s) > 4:
+                    server_urls[s['ilo']]=s
+                pass
+        else:
+            logwriter(log,'Exception - ILO is not reachable: '+s['ilo'])
+    logclose(log)
+    return server_urls
+```
+
+## g﻿et\_server\_data
+
+The get\_server\_data routine is using the obtained server\_urls of one of the monitored server together with the Prometheus user information to retrieve the power and thermal data as well as some general system information data. The captured data is packed into a server\_data directory and returned to the main program.
```

Comment:
I recommend making "the monitored server" pural, since you're mentioning "one of the monitored". So it should read "one of the monitored servers together with"

Again, here, start the paragraph with "The get\_server\_data routine uses the..."